contracts/SolverTrampoline.sol

```diff
@@ -17,17 +23,41 @@ contract SolverTrampoline {
     /// @dev The domain separator for signing EIP-712 settlements.
     bytes32 public immutable domainSeparator;
 
+    /// @dev Nonce by solver address.
+    mapping(address => uint256) public nonces;
+
     constructor(Settlement settlementContract_) {
         settlementContract = settlementContract_;
         solverAuthenticator = settlementContract_.authenticator();
-        
+
         domainSeparator = keccak256(abi.encode(
             keccak256("EIP712Domain(uint256 chainId,address verifyingContract)"),
             block.chainid,
             address(this)
         ));
     }
 
+    /// @dev Executes a settlement on behalf of a solver.
+    function settle(bytes calldata settlement, uint256 nonce, bytes32 r, bytes32 s, uint8 v) external {
+        bytes32 messageDigest = settlementMessage(settlement, nonce);
+        address solver = ecrecover(messageDigest, v, r, s);
+        if (solver == address(0) || !solverAuthenticator.isSolver(solver)) {
+            revert Unauthorized();
+        }
+        if (nonce != nonces[solver]) {
+            revert InvalidNonce();
+        }
+
+        nonces[solver] = nonce + 1;
+        (bool success, bytes memory data) = address(settlementContract).call(settlement);
+        if (!success) {
+            // Propagate the revert error.
+            assembly {
+                revert(add(data, 32), mload(data))
```

Comment:
Do we want to revert if the call reverts (i.e., not consume the nonce)?

Advantages:
- the call to this contract "looks similar" to a call from an EOA and yields the expected error as usual on failure (probably less code changes needed to accommodate the trampoline)

Disadvantages:
- calls can be retried in case of failure (price movements might make it advantageous to execute this transaction later, for example because of internalized trades; all information needed to do so is publicly available on the blockchain for everyone to use)

Dubious points:
- If two calls are sent by the same solver with nonces `nonce` and `nonce + 1`:
   - In the non-reverting case, the second call isn't blocked by the first failing. However, anyone can make the first fail in order to execute only the second in a single Ethereum transaction. This means the second cannot assume the state changes of the first have been applied (otoh, if they were related they would be in the same settlement?).
   - In the reverting case, the second transaction won't be executed until the first transaction is replaced (or cancelled, which I imagine will be implemented at some point).

For comparison with other projects: the Safe contract does not revert in this case.
My personal preference would be not reverting on failure, but I don't have a notion of how complicated it would be to tackle in the services.

Reply:
https://github.com/cowprotocol/solver-trampoline/pull/4#discussion_r1053506414